File: solutions/python/pig-latin/pig_latin.py
vowels = ["a", "e", "i", "o", "u"]
consonants = [c for c in "abcdefghijklmnopqrstuvwxyz" if c not in vowels]

def initial_consonants(word):

    result = ""

    for i, ch in enumerate(word):
        if i < 1 and ch in consonants:
            result += ch
        elif ch in consonants and ch != "y":
            result += ch
        else:
            return result

    return result

def translate_word(word):

    ic = initial_consonants(word)

    if word[0] in vowels or word[:2] in ["xr", "yt"]:
        return word + "ay"

    if ic and ic[-1] == "q" and len(word) > len(ic) and word[len(ic)] == "u":
        return word[len(ic) + 1:] + ic + "uay"

    if ic and len(word) > len(ic) and word[len(ic)] == "y":
        return word[len(ic):] + ic + "ay"

    if word[0] in consonants:
        return word[len(ic):] + ic + "ay"

    return word

def translate(text):
    return " ".join(translate_word(w) for w in text.split())

File: solutions/julia/pig-latin/pig-latin.jl
using Base.Iterators

const vowels = ['a', 'e', 'i', 'o', 'u']

const consonants = [c for c in 'a':'z' if c ∉ vowels]

function initial_consonants(word)

    result = ""

    for i in eachindex(word)
        if i < 2 && word[i] ∈ consonants
            result *= word[i]
        elseif word[i] ∈ consonants && word[i] != 'y'
            result *= word[i]
        else
            return result
        end
    end

    return result

end

function translate_word(word)

    ic = initial_consonants(word)

    if word[1] ∈ vowels || (word[1:2] ∈ ["xr", "yt"])
        return word * "ay"
    elseif ic[end] == 'q' && length(word) > length(ic) && word[length(ic) + 1] == 'u'
        return word[length(ic) + 2:end] * ic * "uay"
    elseif length(ic) > 0 && length(word) > length(ic) && word[length(ic) + 1] == 'y'
        return word[length(ic) + 1 : end] * ic[1:end] * "ay"
    elseif word[1] ∈ consonants
        return word[length(ic) + 1 : end] * word[1 : length(ic)] * "ay"
    else
        return word
    end
end

function translate(phrase)
    return join([translate_word(word) for word in split(phrase)], " ")
end
